### PR TITLE
PEOPLE: Data quality checker

### LINKED ISSUE #795: PEOPLE: Überprüfung Datenqualität
Als Mitgliederverwaltung SAC möchte ich, das beim Mutieren von Personendaten die Datenqualität überprüft wird.

Diese Prüfungen werden in einer eigenen Domainklasse implementiert (z.B. PersonDataQualityChecker), so dass sie auch anderweitig angestossen werden können. Es sollen also nicht ActiveRecord Validations verwendet werden.

Beim Speichern einer Person (nur falls eines der unten erwähnten Attribute geändert hat => Liste aller geprüften Attribute als Konstante in der neuen Domainklasse führen; after_save Callback) sowie beim Erstellen oder Löschen einer `SacCas::Role::MitgliedCommon` (Model after_create / after_destroy Callbacks) soll die Domainklasse angestossen werden.

Für jede Prüfung, deren Bedingung erfüllt ist, wir ein Eintrag in der Tabelle `Person::DataQualityIssue` erstellt, falls dieser nicht bereit existiert. Der Key für die Fehlermeldung, das Attribut und die Stufe sind in der untenstehenden Tabelle ersichtlich. Falls die Bedingung nicht erfüllt ist, wird ein allfällig existierender Eintrag gelöscht.

### Prüfungen

|*Fehlermeldung (key)*|*Attribut (attr)*| *Bedingung*|*Stufe (severity)*|
|---|---|---| ---|
|ist leer| email | MV Rechnungsempfänger && Person#email.blank?|warning |
|ist leer| street | MV Rechnungsempfänger && Person#street.blank? | error |
|ist leer| zip_code | MV Rechnungsempfänger && Person#zip_code.blank? | error |
|ist leer| town | MV Rechnungsempfänger &&  Person#town.blank? | error |
|ist leer| first_name | !Person.company? &&  Person#first_name.blank? | error |
|ist leer| last_name |  !Person.company? && Person#last_name.blank? | error |
|ist leer| company_name |  Person.company? && Person#company_name.blank? | warning |
|ist leer| phone_numbers |  MV Rechnungsempfänger  && Person#phone_numbers.blank? | warning |
|ist leer| birthday |  Rolle `SacCas::STAMMSEKTION_ROLES` vorhanden && Person#birthday.blank? | error |
| liegt weniger als 6 Jahre vor dem SAC-Eintritt|  birthday |  Rolle  `SacCas::STAMMSEKTION_ROLES`  vorhanden  && Person#birthday.present? && Person#birthday > Role#created_at - 6.years | warning |

Die Bedingung "MV Rechnungsempfänger" entspricht der Methode `Person#sac_membership_invoice?`. Bei den Prüfungen sollten keine spezifischen Datenbankqueries abgesetzt werden, sondern mit der (potentiell schon vorgängig bereits geladenen) `roles` Association auf der Person gearbeitet werden.

Wenn alle Prüfungen durchgelaufen sind, wird auf der Person das Attribut `data_quality` auf die höchste gefundene Stufe gesetzt. Falls keine Einträge vorhanden sind, wird der Wert auf `ok` gesetzt.

# ToDo

- [x] Domainklasse erstellen
- [x] Model anpassen
- [x] Specs schreiben
- [ ]  Kunde wegen Übersetzungen informieren

### PR DESCRIPTION
**Pull request by hunchr**
Fixes #795

### CODE REVIEW

File: app/domain/people/data_quality_checker.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+#  Copyright (c) 2024, Schweizer Alpen-Club. This file is part of
+#  hitobito_sac_cas and licensed under the Affero General Public License version 3
+#  or later. See the COPYING file at the top-level directory or at
+#  https://github.com/hitobito/hitobito_sac_cas
+
+class People::DataQualityChecker
```
Comment (njaeggi) [inline app/domain/people/data_quality_checker.rb:8]: Would be cleaner if we defined the attributes to check like this

```
ATTRIBUTES_TO_CHECK = {
    "street" => { severity: "error", key: "empty", invalid: @person.street.blank, perform_check:  @person.sac_membership_invoice?}
  }
```
  
  and then call check like this
  
```
  ATTRIBUTES_TO_CHECK.each do |field, data|
    check(data[:severity], field, data[:message], invalid: data[:invalid]) if data[:perform_check]
  end
  
```

File: app/models/sac_cas/person.rb [hunk 2]
```diff
@@ -99,4 +100,12 @@ def backoffice?
   def sac_membership
     @sac_membership ||= People::SacMembership.new(self)
   end
+
+  private
+
+  def check_data_quality
+    return if (People::DataQualityChecker::ATTRIBUTES_TO_CHECK & saved_changes.keys).empty?
+
+    People::DataQualityChecker.new(self)
```
Comment (njaeggi) [inline app/models/sac_cas/person.rb:109]: Add a method to call, instead of calling the checks on initialization, for example `check_data_quality`

**Review by njaeggi — CHANGES_REQUESTED** [review]
Looks pretty good already :rocket:

File: app/domain/people/data_quality_checker.rb (hunk 1 above)
Comment (hunchr) [reply app/domain/people/data_quality_checker.rb:8]: `ATTRIBUTES_TO_CHECK` cannot be a hash because `:birthday` must be checked twice

**Review by njaeggi — APPROVED** [review]
Nice!

File: app/domain/people/data_quality_checker.rb (hunk 1 above)
Comment (njaeggi) [reply app/domain/people/data_quality_checker.rb:8]: Didn't test my solution, was only a quick suggestion, but your solution now is good :+1:

**Pull request merged**